src/utils/style-utils.js

```diff
@@ -114,3 +87,49 @@ function updateStyleSource(map, update) {
   map.removeSource(update.id);
   map.addSource(update.id, newSource);
 }
+
+function applySourceLayerChanges(map, nextStyle, sourcesDiff, layersDiff) {
+// TODO: It's rather difficult to determine style diffing in the presence
+  // of refs. For now, if any style update has a ref, fallback to no diffing.
+  // We can come back to this case if there's a solid usecase.
+  if (layersDiff.updates.some(node => node.layer.get('ref'))) {
+    map.setStyle(nextStyle.toJS());
+    return;
+  }
+
+  for (const enter of sourcesDiff.enter) {
+    map.addSource(enter.id, enter.source.toJS());
+  }
+  for (const update of sourcesDiff.update) {
+    updateStyleSource(map, update);
+  }
+
+  for (const exit of layersDiff.exiting) {
+    if (map.style.getLayer(exit.id)) {
+      map.removeLayer(exit.id);
+    }
+  }
+  for (const update of layersDiff.updates) {
+    if (!update.enter) {
+      // This is an old layer that needs to be updated. Remove the old layer
+      // with the same id and add it back again.
+      map.removeLayer(update.id);
+    }
+    map.addLayer(update.layer.toJS(), update.before);
+  }
+
+  for (const exit of sourcesDiff.exit) {
+    map.removeSource(exit.id);
+  }
+}
+
+/* eslint-disable max-len */
+function checkForEqualLayerSourceChanges(sourceExit, nextLayers, callback) {
```

Comment:
Any particular reason to use a callback here?

Reply:
Sorry for the delayed response! I left it there in-case the check needed to be used in the future to for a different reason. I don't see an impact to dropping the callback and either placing `applySourceLayerChanges` after the check inside `checkForEqualLayerSourceChanges` or making `checkForEqualLayerSourceChanges` sync and just returning a flag if things are good.